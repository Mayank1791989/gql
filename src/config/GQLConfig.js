/* @flow */
/* eslint-disable no-use-before-define */
import findConfig from 'find-config';
import JSON5 from 'json5';
import fs from 'fs';
import minimatch from 'minimatch';
import parseGlob from 'parse-glob';
import path from 'path';
import _uniq from 'lodash/uniq';

type Options = {
  cwd?: string
};

const CONFIG_FILE_NAME = '.gqlconfig';

export type QueryParser = (
  'QueryParser'
  | ['EmbeddedQueryParser', { startTag: string, endTag: string }]
);
export type Globs = string | Array<string>;
export type FileMatchConfig = Globs | { include: Globs, ignore?: Globs };
export type ValidateConfig = {
  extends: string,
  rules?: {
    [ruleName: string]: 'off' | 'warn' | 'error',
  },
};
export type QueryConfig = {
  match: FileMatchConfig,
  parser: QueryParser,
  isRelay?: boolean,
  validate?: ValidateConfig,
};

type GQLConfigFile = {
  schema: {
    files: FileMatchConfig,
    validate?: ValidateConfig,
  },
  query?: { // query optional
    files: Array<QueryConfig>,
  },
};

export type { Options as GQLConfigOptions };

class GQLConfig {
  _path: string;
  _configObj: GQLConfigFile;
  _dir: string;

  constructor(options: ?Options) {
    const result = findConfig.obj(CONFIG_FILE_NAME, options);
    if (!result) {
      // throw error .gqlConfig not found
      throw new Error(`Could not find a ${CONFIG_FILE_NAME} file. Create a ${CONFIG_FILE_NAME} file in project root directory.`);
    }

    this._dir = result.dir;
    this._path = result.path;
    this._configObj = this._readConfig(result.path);
  }

  getDir() {
    return this._dir;
  }

  getPath() {
    return this._path;
  }

  getFileExtensions(): Array<string> {
    const { schema, query } = this._configObj;
    const extensions = [];

    // schema
    extensions.push(...extractExtensions(schema.files));

    // query
    if (query) {
      query.files.forEach(({ match }) => {
        extensions.push(...extractExtensions(match));
      });
    }

    return _uniq(extensions);
  }

  getSchema() {
    return this._configObj.schema;
  }

  getQuery() {
    return this._configObj.query;
  }

  match(filePath: string): ?{ type: 'query' | 'schema', opts: Object } {
    const { schema, query } = this._configObj;
    const filePathRelative = path.relative(this.getDir(), filePath);

    // test schema
    if (matcher(filePathRelative, schema.files)) {
      return { type: 'schema', opts: {} };
    }

    // test query
    if (!query) { return null; }

    const { files } = query;
    const opts = files.find(({ match }) => matcher(filePathRelative, match));
    if (opts) {
      return { type: 'query', opts };
    }

    return null;
  }

  _readConfig(filePath: string) { // eslint-disable-line class-methods-use-this
    const fileData = fs.readFileSync(filePath, 'utf8');
    try {
      return (JSON5.parse(fileData): GQLConfigFile); // using flow-runtime it will be validated also
    } catch (err) {
      throw new Error(`\nError parsing ${CONFIG_FILE_NAME}. \n\n${err.message}`);
    }
  }
}

function extractExtensions(match: FileMatchConfig): Array<string> {
  if (typeof match === 'string') {
    return [getExtFromGlob(match)];
  }

  if (Array.isArray(match)) {
    return match.map(getExtFromGlob);
  }

  // object { match, exclude }
  return extractExtensions(match.include);
}

function getExtFromGlob(glob: string): string {
  return parseGlob(glob).path.ext; // without dot
}

function matchGlob(filePath, globs: Globs): boolean {
  if (typeof globs === 'string') {
    return minimatch(filePath, globs);
  }

  // matches any
  return Boolean(globs.find(glob => minimatch(filePath, glob)));
}

function matcher(filePath, match: FileMatchConfig): boolean {
  if (typeof match === 'string' || Array.isArray(match)) {
    return matchGlob(filePath, match);
  }

  const _match = matchGlob(filePath, match.include);
  if (!_match) { return false; }

  const _matchesIgnore = match.ignore ? matchGlob(filePath, match.ignore) : false;

  return _match && !_matchesIgnore;
}

export default GQLConfig;
