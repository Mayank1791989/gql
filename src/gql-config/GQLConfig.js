/* @flow */
import findConfigFile from './findConfigFile';
import readConfigFile from './readConfigFile';
import resolveConfigFile from './resolveConfigFile';
import extractExtensions from './extractExtensions';
import getConfigForFile, {
  type SchemaFileConfig,
  type QueryFileConfig,
} from './getConfigForFile';

import _memoize from 'lodash/memoize';

import { type GQLConfigFileResolved } from './types';
import {
  type ParseOptions as GraphQLParseOptions,
  type BuildSchemaOptions as GraphQLBuildSchemaOptions,
} from 'graphql';

class GQLConfig {
  _dir: string;
  _path: string;
  _configFileResolved: GQLConfigFileResolved;

  constructor(options: { configDir: string }) {
    const result = findConfigFile(options.configDir);
    this._dir = result.dir;
    this._path = result.path;
    this._configFileResolved = resolveConfigFile(
      readConfigFile(result.path),
      result.dir,
    );
    // console.log(JSON.stringify(this._configFileResolved, null, 2));
  }

  // return directory containing config file
  getDir(): string {
    return this._dir;
  }

  // return path of config file
  getPath(): string {
    return this._path;
  }

  // return all file extensions in config file
  getFileExtensions: () => Array<string> = _memoize(() =>
    extractExtensions(this._configFileResolved),
  );

  // return schema config
  getSchemaConfig(): $PropertyType<GQLConfigFileResolved, 'schema'> {
    return this._configFileResolved.schema;
  }

  // return query config
  getQueryConfig(): $PropertyType<GQLConfigFileResolved, 'query'> {
    return this._configFileResolved.query;
  }

  // returns config for given file path
  // file can be schema file or any query file
  getFileConfig(filePath: string): SchemaFileConfig | QueryFileConfig | null {
    return getConfigForFile(filePath, this._configFileResolved, this.getDir());
  }

  getGraphQLParseOptions(): GraphQLParseOptions {
    const { graphQLOptions } = this.getSchemaConfig();
    const options = {};

    if (
      graphQLOptions &&
      typeof graphQLOptions.allowLegacySDLEmptyFields === 'boolean'
    ) {
      options.allowLegacySDLEmptyFields =
        graphQLOptions.allowLegacySDLEmptyFields;
    }

    if (
      graphQLOptions &&
      typeof graphQLOptions.allowLegacySDLImplementsInterface === 'boolean'
    ) {
      options.allowLegacySDLImplementsInterfaces =
        graphQLOptions.allowLegacySDLImplementsInterface;
    }

    return options;
  }

  getGraphQLBuildSchemaOptions(): GraphQLBuildSchemaOptions {
    const { graphQLOptions } = this.getSchemaConfig();
    const options = {};
    if (
      graphQLOptions &&
      typeof graphQLOptions.commentDescriptions === 'boolean'
    ) {
      options.commentDescriptions = graphQLOptions.commentDescriptions;
    }
    return options;
  }
}

export default GQLConfig;
