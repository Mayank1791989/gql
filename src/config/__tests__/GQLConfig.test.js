/* @flow */
import GQLConfig from '../GQLConfig';
import path from 'path';

const fixture = (name) => path.join(__dirname, 'fixtures', name);

test('missing-gqlConfig: Report error if gqlconfig not found', () => {
  expect(() => {
    const gqlConfig = new GQLConfig({ cwd: fixture('no-gqlconfig') }); // eslint-disable-line
  }).toThrowErrorMatchingSnapshot();
});

test('invalid-gqlconfig: validate .gqlconfig file content', () => {
  expect(() => {
    const gqlConfig = new GQLConfig({ cwd: fixture('invalid-gqlconfig') }); // eslint-disable-line
  }).toThrowErrorMatchingSnapshot();
});

test('only-schema-gqlconfig: allow only schema', () => {
  expect(() => {
    const gqlConfig = new GQLConfig({ cwd: fixture('only-schema-gqlconfig') }); // eslint-disable-line
  }).not.toThrowError();
});

describe('config.match', () => {
  const gqlConfig = new GQLConfig({ cwd: fixture('valid-gqlconfig') });
  const genPath = (file) => fixture(`valid-gqlconfig/${file}`);

  test('match schema files', () => {
    expect(
      gqlConfig.match(genPath('schema/player.gql')),
    ).toMatchSnapshot();
  });

  test('match query files', () => {
    expect(
      gqlConfig.match(genPath('query/player_query.gql')),
    ).toMatchSnapshot();
  });

  test('dont match ignored files', () => {
    expect(
      gqlConfig.match(genPath('query/ignore/player_query.js')),
    ).toBeNull();
  });
});
