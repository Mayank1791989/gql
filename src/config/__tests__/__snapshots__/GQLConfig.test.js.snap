exports[`config.match match query files 1`] = `
Object {
  "opts": Object {
    "match": "query/**/*.gql",
    "parser": "QueryParser",
  },
  "type": "query",
}
`;

exports[`config.match match schema files 1`] = `
Object {
  "opts": Object {},
  "type": "schema",
}
`;

exports[`test invalid-gqlconfig: validate .gqlconfig file content 1`] = `
"
Error parsing .gqlconfig. 

GQLConfigFile.schema.files must be one of: Globs | {
  include: Globs;
  ignore?: Globs;
}

Expected: Globs | {
  include: Globs;
  ignore?: Globs;
}

Actual: {
  test: string;
}
"
`;

exports[`test missing-gqlConfig: Report error if gqlconfig not found 1`] = `"Could not find a .gqlconfig file. Create a .gqlconfig file in project root directory."`;
